cmd/relay-gossip/client/client.go

```diff
@@ -2,123 +2,73 @@ package client
+	dlog "github.com/drand/drand/log"
 	pubsub "github.com/libp2p/go-libp2p-pubsub"
 	"golang.org/x/xerrors"
 )
 
-var (
-	log = logging.Logger("drand-client")
-)
-
-type Client struct {
-	cancel func()
-	latest uint64
+// ErrNotAvailable is returned when Get is called using a basic client with no
+// HTTP API endpoints configured and value was not cached via gossip.
+var ErrNotAvailable = fmt.Errorf("not available")
+
+// Options are configuration options for the drand libp2p pubsub client
+type Options struct {
+	// HTTPEndpoints are drand HTTP API URL(s) to use incase of gossipsub failure.
+	HTTPEndpoints []string
+	// FailoverGracePeriod is the period before the failover HTTP API is used when watching for randomness.
+	FailoverGracePeriod time.Duration
+	// Logger is a custom logger to use with the client
+	Logger dlog.Logger
+}
 
-	subs struct {
-		sync.Mutex
-		M map[*int]chan drand.PublicRandResponse
-	}
+type basicClient struct {
+	group       *key.Group
+	getNotifier dclient.GetNotifierFunc
 }
 
-// NewWtihPubsub creates a gossip randomness client.
-func NewWithPubsub(ps *pubsub.PubSub, networkName string) (*Client, error) {
-	t, err := ps.Join(lp2p.PubSubTopic(networkName))
+// NewWithPubsub creates a new drand client that uses pubsub to receive randomness updates.
+func NewWithPubsub(ps *pubsub.PubSub, group *key.Group, options Options) (dclient.Client, error) {
+	t, err := ps.Join(lp2p.PubSubTopic(string(group.Hash())))
 	if err != nil {
 		return nil, xerrors.Errorf("joining pubsub: %w", err)
 	}
-	s, err := t.Subscribe()
-	if err != nil {
-		return nil, xerrors.Errorf("subscribe: %w", err)
-	}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	c := &Client{
-		cancel: cancel,
+	log := options.Logger
+	if log == nil {
+		log = dlog.DefaultLogger
 	}
-	c.subs.M = make(map[*int]chan drand.PublicRandResponse)
-
-	go func() {
-		for {
-			msg, err := s.Next(ctx)
-			if ctx.Err() != nil {
-				c.subs.Lock()
-				for _, ch := range c.subs.M {
-					close(ch)
-				}
-				c.subs.M = make(map[*int]chan drand.PublicRandResponse)
-				c.subs.Unlock()
-				t.Close()
-				s.Cancel()
-				return
-			}
-			if err != nil {
-				log.Warnf("topic.Next error: %+v", err)
-				continue
-			}
-			var rand drand.PublicRandResponse
-			err = proto.Unmarshal(msg.Data, &rand)
-			if err != nil {
-				log.Warnf("unmarshaling randomness: %+v", err)
-				continue
-			}
-
-			// TODO: verification, need to pass drand network public key in
 
-			if c.latest >= rand.Round {
-				continue
-			}
-			c.latest = rand.Round
-
-			c.subs.Lock()
-			for _, ch := range c.subs.M {
-				select {
-				case ch <- rand:
-				default:
-					log.Warn("randomness notification dropped due to a full channel")
-				}
-			}
-			c.subs.Unlock()
-
-		}
-	}()
+	if len(options.HTTPEndpoints) == 0 {
+		bc := basicClient{group: group, getNotifier: NewNotifier(t, log)}
+		return dclient.NewCachingClient(&bc, 32, log)
+	}
 
-	return c, nil
+	return dclient.New(
```

Comment:
Since you are using dclient.Client inside NewWithPubSub, effectively NewWithPubSub becomes the new official top-level client (it is the only one that supports both REST and gossipsub, each optionally).

This is not very self evident from the code. Most folks will reach out for drand/client, when they should be using drand/cmd/relay-gossip/client. We need to refactor things a bit? At minimum, I would move the creation of this client to drand/client? Other alternatives?

Reply:
This confusion is exacerbated from the cosmetics of the fact that dclient.New uses the proper With...() options semantics. Whereas this one uses the more "casual" configure with a structure (type Options) approach.

Reply:
> This is not very self evident from the code. Most folks will reach out for drand/client, when they should be using drand/cmd/relay-gossip/client. We need to refactor things a bit?

I think most folks will reach out for `drand/client` and expect a HTTP only client. My understanding is that the gossip client is an alternative that can be used should the consumer happen to be using the libp2p stack.

`drand/cmd/relay-gossip` has a separate `go.mod` to avoid regular users of drand having to pull in many dependencies they will never use.

I'm happy to switch to a libp2p oriented implementation but this feels like the right arrangement of things to me.

Reply:
in the [design](https://docs.google.com/document/d/1UFKqF-CTHjPd4NjHjmQ6ljgLtqoh-huzsUskYFaA-NM/edit?ts=5ec4da8d) i sketched, we had `drand/client` as the top level client. I would continue to prefer that to be the place you go to get "a client" regardless of if it's http only, a hybrid, or gossip sub.

That way there's one place doing the composition of caching / fallback / etc, and it's easier to ensure we're doing that consistently and know where/how to fix bugs.

I expect a consumer using gossip will be pulling in a `drand/client/gossip` or similar package that exposes the config.Option where you pass in the `libp2p` types / dependencies needed for the gossip creation, and the core `client` package can avoid requiring those to be direct dependencies with an interface, as you've already done with the `Notifier`.

Reply:
SGTM